File: Vanta Speech iOS/Vanta Speech/Core/EAS/API/EASXMLParser.swift
import Foundation

/// Parser for EAS XML responses
final class EASXMLParser: NSObject {

    private let xml: String
    private var parser: XMLParser?

    // Parsing state
    private var currentElement = ""
    private var currentText = ""

    // FolderSync parsing
    private var folderSyncKey = ""
    private var folderSyncStatus = 0
    private var folders: [EASFolder] = []
    private var currentFolder: FolderBuilder?

    // Sync parsing
    private var syncKey = ""
    private var syncStatus = 0
    private var events: [EASCalendarEvent] = []
    private var currentEvent: EventBuilder?
    private var currentAttendee: AttendeeBuilder?
    private var moreAvailable = false

    // Namespace tracking
    private var inFolder = false
    private var inEvent = false
    private var inAttendee = false
    private var inBody = false
    private var inRecurrence = false
    private var currentRecurrence: RecurrenceBuilder?

    // Response type tracking
    private var isFolderSyncResponse = false
    private var isSyncResponse = false

    init(xml: String) {
        self.xml = xml
        super.init()
    }

    // MARK: - Public API

    func parseFolderSync() throws -> FolderSyncResponse {
        guard let data = xml.data(using: .utf8) else {
            throw EASError.parseError("Invalid XML encoding")
        }

        // Mark that we're parsing a FolderSync response
        isFolderSyncResponse = true
        isSyncResponse = false

        parser = XMLParser(data: data)
        parser?.delegate = self
        parser?.parse()

        if let error = parser?.parserError {
            throw EASError.parseError(error.localizedDescription)
        }

        return FolderSyncResponse(
            syncKey: folderSyncKey,
            folders: folders,
            status: folderSyncStatus
        )
    }

    func parseSync() throws -> SyncResponse {
        guard let data = xml.data(using: .utf8) else {
            throw EASError.parseError("Invalid XML encoding")
        }

        // Mark that we're parsing a Sync response
        isFolderSyncResponse = false
        isSyncResponse = true

        parser = XMLParser(data: data)
        parser?.delegate = self
        parser?.parse()

        if let error = parser?.parserError {
            throw EASError.parseError(error.localizedDescription)
        }

        return SyncResponse(
            syncKey: syncKey,
            events: events,
            status: syncStatus,
            moreAvailable: moreAvailable
        )
    }
}

// MARK: - XMLParserDelegate

extension EASXMLParser: XMLParserDelegate {

    func parser(
        _ parser: XMLParser,
        didStartElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?,
        attributes attributeDict: [String: String] = [:]
    ) {
        currentElement = elementName
        currentText = ""

        // Debug: log element names related to attendees
        if elementName.lowercased().contains("attend") || elementName == "Email" || elementName == "Name" {
            print("[EAS Parser] START element: \(elementName), inEvent=\(inEvent), inAttendee=\(inAttendee)")
        }

        switch elementName {
        case "Add", "Update":
            if currentFolder == nil && currentEvent == nil {
                // Determine type based on response type, not folder count
                if isFolderSyncResponse {
                    // In FolderSync response, Add means folder
                    currentFolder = FolderBuilder()
                    inFolder = true
                } else if isSyncResponse {
                    // In Sync response, Add means calendar item
                    currentEvent = EventBuilder()
                    inEvent = true
                }
            }
        case "Attendee":
            currentAttendee = AttendeeBuilder()
            inAttendee = true
        case "Body":
            inBody = true
        case "Recurrence":
            currentRecurrence = RecurrenceBuilder()
            inRecurrence = true
        case "MoreAvailable":
            moreAvailable = true
        default:
            break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        currentText += string
    }

    func parser(
        _ parser: XMLParser,
        didEndElement elementName: String,
        namespaceURI: String?,
        qualifiedName qName: String?
    ) {
        let text = currentText.trimmingCharacters(in: .whitespacesAndNewlines)

        // Handle folder parsing
        if inFolder {
            switch elementName {
            case "ServerId":
                currentFolder?.serverId = text
            case "ParentId":
                currentFolder?.parentId = text
            case "DisplayName":
                currentFolder?.displayName = text
            case "Type":
                currentFolder?.type = Int(text) ?? 0
            case "Add", "Update":
                if let folder = currentFolder?.build() {
                    folders.append(folder)
                }
                currentFolder = nil
                inFolder = false
            default:
                break
            }
        }

        // Handle event parsing
        if inEvent {
            switch elementName {
            case "ServerId":
                currentEvent?.id = text
            case "Subject":
                currentEvent?.subject = text
            case "StartTime":
                currentEvent?.startTime = parseDate(text)
            case "EndTime":
                currentEvent?.endTime = parseDate(text)
            case "Location":
                currentEvent?.location = text
            case "AllDayEvent":
                currentEvent?.isAllDay = text == "1"
            case "OrganizerEmail":
                currentEvent?.organizerEmail = text
                print("[EAS Parser] Organizer email: \(text)")
            case "OrganizerName":
                currentEvent?.organizerName = text
                print("[EAS Parser] Organizer name: \(text)")
            case "Data":
                if inBody {
                    currentEvent?.body = text
                    print("[EAS Parser] Body data length: \(text.count)")
                }
            case "Body":
                inBody = false
            case "Add", "Update":
                if let event = currentEvent?.build() {
                    events.append(event)
                }
                currentEvent = nil
                inEvent = false
            default:
                break
            }
        }

        // Handle attendee parsing - MUST check BEFORE event parsing to catch Email/Name inside Attendee
        if inAttendee {
            switch elementName {
            case "Email":
                currentAttendee?.email = text
                print("[EAS Parser] Attendee email set: \(text)")
            case "Name":
                currentAttendee?.name = text
                print("[EAS Parser] Attendee name set: \(text)")
            case "AttendeeType":
                currentAttendee?.type = Int(text) ?? 1
            case "AttendeeStatus":
                currentAttendee?.status = Int(text)
            case "Attendee":
                if let attendee = currentAttendee?.build() {
                    currentEvent?.attendees.append(attendee)
                    print("[EAS Parser] ADDED attendee: \(attendee.name) <\(attendee.email)>")
                } else {
                    print("[EAS Parser] FAILED to build attendee: email=\(currentAttendee?.email ?? "nil"), name=\(currentAttendee?.name ?? "nil")")
                }
                currentAttendee = nil
                inAttendee = false
            default:
                break
            }
        }

        // Handle recurrence parsing
        if inRecurrence {
            switch elementName {
            case "Type":
                currentRecurrence?.type = Int(text)
            case "Interval":
                currentRecurrence?.interval = Int(text)
            case "DayOfWeek":
                currentRecurrence?.dayOfWeek = Int(text)
            case "DayOfMonth":
                currentRecurrence?.dayOfMonth = Int(text)
            case "WeekOfMonth":
                currentRecurrence?.weekOfMonth = Int(text)
            case "MonthOfYear":
                currentRecurrence?.monthOfYear = Int(text)
            case "Until":
                currentRecurrence?.until = parseDate(text)
            case "Occurrences":
                currentRecurrence?.occurrences = Int(text)
            case "Recurrence":
                if let recurrence = currentRecurrence?.build() {
                    currentEvent?.recurrence = recurrence
                    print("[EAS Parser] Recurrence: type=\(recurrence.type.rawValue), interval=\(recurrence.interval), dayOfWeek=\(recurrence.dayOfWeek ?? 0)")
                }
                currentRecurrence = nil
                inRecurrence = false
            default:
                break
            }
        }

        // Handle root elements
        switch elementName {
        case "SyncKey":
            if folderSyncKey.isEmpty {
                folderSyncKey = text
            }
            if syncKey.isEmpty {
                syncKey = text
            }
        case "Status":
            if let status = Int(text) {
                if folderSyncStatus == 0 {
                    folderSyncStatus = status
                }
                if syncStatus == 0 {
                    syncStatus = status
                }
            }
        default:
            break
        }

        currentElement = ""
        currentText = ""
    }

    // MARK: - Helpers

    private func parseDate(_ string: String) -> Date? {
        // EAS uses compact ISO8601 format: 20251107T100000Z (no dashes or colons)
        // Try compact format first (most common in EAS)
        let compactFormatter = DateFormatter()
        compactFormatter.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
        compactFormatter.timeZone = TimeZone(identifier: "UTC")
        compactFormatter.locale = Locale(identifier: "en_US_POSIX")

        if let date = compactFormatter.date(from: string) {
            return date
        }

        // Try standard ISO8601 format
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        if let date = formatter.date(from: string) {
            return date
        }

        // Try without fractional seconds
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: string)
    }
}

// MARK: - Builder Classes

private class FolderBuilder {
    var serverId = ""
    var parentId = "0"
    var displayName = ""
    var type = 0

    func build() -> EASFolder? {
        guard !serverId.isEmpty else { return nil }

        let folderType = EASFolderType(rawValue: type) ?? .unknown

        return EASFolder(
            serverId: serverId,
            parentId: parentId,
            displayName: displayName,
            type: folderType
        )
    }
}

private class EventBuilder {
    var id = ""
    var subject = ""
    var startTime: Date?
    var endTime: Date?
    var location: String?
    var body: String?
    var isAllDay = false
    var attendees: [EASAttendee] = []
    var organizerEmail: String?
    var organizerName: String?
    var recurrence: EASRecurrence?

    func build() -> EASCalendarEvent? {
        guard !id.isEmpty, let start = startTime else {
            print("[EAS Parser] Event build failed: id='\(id)', startTime=\(String(describing: startTime)), endTime=\(String(describing: endTime))")
            return nil
        }

        // If EndTime is missing, default to StartTime + 1 hour (or end of day for all-day events)
        let end: Date
        if let endTime = endTime {
            end = endTime
        } else if isAllDay {
            // All-day event: end at start of next day
            end = Calendar.current.date(byAdding: .day, value: 1, to: start) ?? start
        } else {
            // Regular event: default to 1 hour duration
            end = Calendar.current.date(byAdding: .hour, value: 1, to: start) ?? start
        }

        // Build organizer from email/name if available
        var organizer: EASAttendee?
        if let email = organizerEmail, !email.isEmpty {
            organizer = EASAttendee(
                email: email,
                name: organizerName ?? email,
                type: .required,
                status: nil
            )
        }

        print("[EAS Parser] Built event: '\(subject)' from \(start) to \(end), organizer: \(organizer?.email ?? "none"), attendees: \(attendees.count), recurring: \(recurrence != nil)")

        return EASCalendarEvent(
            id: id,
            subject: subject.isEmpty ? "Untitled" : subject,
            startTime: start,
            endTime: end,
            location: location,
            body: body,
            organizer: organizer,
            attendees: attendees,
            isAllDay: isAllDay,
            recurrence: recurrence
        )
    }
}

private class AttendeeBuilder {
    var email = ""
    var name = ""
    var type = 1
    var status: Int?

    func build() -> EASAttendee? {
        guard !email.isEmpty else { return nil }

        return EASAttendee(
            email: email,
            name: name.isEmpty ? email : name,
            type: EASAttendeeType(rawValue: type) ?? .required,
            status: status.flatMap { EASResponseStatus(rawValue: $0) }
        )
    }
}

private class RecurrenceBuilder {
    var type: Int?
    var interval: Int?
    var dayOfWeek: Int?
    var dayOfMonth: Int?
    var weekOfMonth: Int?
    var monthOfYear: Int?
    var until: Date?
    var occurrences: Int?

    func build() -> EASRecurrence? {
        guard let typeValue = type,
              let recurrenceType = RecurrenceType(rawValue: typeValue) else {
            return nil
        }

        return EASRecurrence(
            type: recurrenceType,
            interval: interval ?? 1,
            dayOfWeek: dayOfWeek,
            dayOfMonth: dayOfMonth,
            until: until
        )
    }
}

File: Vanta Sppech Android/app/src/main/java/com/vanta/speech/core/eas/api/EASXMLParser.kt
package com.vanta.speech.core.eas.api

import com.vanta.speech.core.eas.model.*
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.StringReader
import java.text.SimpleDateFormat
import java.util.*

/**
 * Parser for EAS XML responses
 */
class EASXMLParser {

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    private val dateFormatAlt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    /**
     * Parse FolderSync response
     */
    fun parseFolderSync(xml: String): FolderSyncResponse {
        val parser = createParser(xml)
        var syncKey = ""
        var status = 0
        val folders = mutableListOf<EASFolder>()

        var currentFolder: FolderBuilder? = null
        var inAdd = false

        while (parser.eventType != XmlPullParser.END_DOCUMENT) {
            when (parser.eventType) {
                XmlPullParser.START_TAG -> {
                    when (parser.name) {
                        "Add", "Update" -> {
                            currentFolder = FolderBuilder()
                            inAdd = true
                        }
                    }
                }
                XmlPullParser.TEXT -> {
                    val text = parser.text?.trim() ?: ""
                    if (text.isNotEmpty()) {
                        when (parser.name) {
                            "SyncKey" -> if (syncKey.isEmpty()) syncKey = text
                            "Status" -> if (status == 0) status = text.toIntOrNull() ?: 0
                        }
                        if (inAdd && currentFolder != null) {
                            when (parser.name) {
                                "ServerId" -> currentFolder.serverId = text
                                "ParentId" -> currentFolder.parentId = text
                                "DisplayName" -> currentFolder.displayName = text
                                "Type" -> currentFolder.type = text.toIntOrNull() ?: 0
                            }
                        }
                    }
                }
                XmlPullParser.END_TAG -> {
                    when (parser.name) {
                        "Add", "Update" -> {
                            currentFolder?.build()?.let { folders.add(it) }
                            currentFolder = null
                            inAdd = false
                        }
                    }
                }
            }
            parser.next()
        }

        return FolderSyncResponse(syncKey, folders, status)
    }

    /**
     * Parse Sync response
     */
    fun parseSync(xml: String): SyncResponse {
        val parser = createParser(xml)
        var syncKey = ""
        var status = 0
        val events = mutableListOf<EASCalendarEvent>()
        var moreAvailable = false

        var currentEvent: EventBuilder? = null
        var currentAttendee: AttendeeBuilder? = null
        var currentException: ExceptionBuilder? = null
        var inAdd = false
        var inAttendee = false
        var inBody = false
        var inExceptions = false
        var inException = false
        var currentTag = ""

        while (parser.eventType != XmlPullParser.END_DOCUMENT) {
            when (parser.eventType) {
                XmlPullParser.START_TAG -> {
                    currentTag = parser.name
                    when (parser.name) {
                        "Add", "Change" -> {
                            currentEvent = EventBuilder()
                            inAdd = true
                        }
                        "Attendee" -> {
                            currentAttendee = AttendeeBuilder()
                            inAttendee = true
                        }
                        "Body" -> inBody = true
                        "Exceptions" -> inExceptions = true
                        "Exception" -> {
                            if (inExceptions) {
                                currentException = ExceptionBuilder()
                                inException = true
                            }
                        }
                        "MoreAvailable" -> moreAvailable = true
                    }
                }
                XmlPullParser.TEXT -> {
                    val text = parser.text?.trim() ?: ""
                    if (text.isNotEmpty()) {
                        when (currentTag) {
                            "SyncKey" -> if (syncKey.isEmpty()) syncKey = text
                            "Status" -> if (status == 0) status = text.toIntOrNull() ?: 0
                        }

                        if (inAdd && currentEvent != null) {
                            when (currentTag) {
                                "ServerId" -> currentEvent.id = text
                                "Subject" -> currentEvent.subject = text
                                "StartTime" -> currentEvent.startTime = parseDate(text)
                                "EndTime" -> currentEvent.endTime = parseDate(text)
                                "Location" -> currentEvent.location = text
                                "AllDayEvent" -> currentEvent.isAllDay = text == "1"
                                "Data" -> if (inBody) currentEvent.body = text
                            }
                        }

                        if (inAttendee && currentAttendee != null) {
                            when (currentTag) {
                                "Email" -> currentAttendee.email = text
                                "Name" -> currentAttendee.name = text
                                "AttendeeType" -> currentAttendee.type = text.toIntOrNull() ?: 1
                                "AttendeeStatus" -> currentAttendee.status = text.toIntOrNull()
                            }
                        }

                        if (inException && currentException != null) {
                            when (currentTag) {
                                "ExceptionStartTime" -> currentException.originalStartTime = parseDate(text)
                                "StartTime" -> currentException.startTime = parseDate(text)
                                "EndTime" -> currentException.endTime = parseDate(text)
                                "Subject" -> currentException.subject = text
                                "Location" -> currentException.location = text
                                "Deleted" -> currentException.isDeleted = text == "1"
                            }
                        }
                    }
                }
                XmlPullParser.END_TAG -> {
                    when (parser.name) {
                        "Add", "Change" -> {
                            currentEvent?.build()?.let { events.add(it) }
                            currentEvent = null
                            inAdd = false
                        }
                        "Attendee" -> {
                            currentAttendee?.build()?.let { currentEvent?.attendees?.add(it) }
                            currentAttendee = null
                            inAttendee = false
                        }
                        "Exception" -> {
                            currentException?.build()?.let { currentEvent?.exceptions?.add(it) }
                            currentException = null
                            inException = false
                        }
                        "Exceptions" -> inExceptions = false
                        "Body" -> inBody = false
                    }
                    currentTag = ""
                }
            }
            parser.next()
        }

        return SyncResponse(syncKey, events, status, moreAvailable)
    }

    private fun createParser(xml: String): XmlPullParser {
        val factory = XmlPullParserFactory.newInstance()
        factory.isNamespaceAware = true
        val parser = factory.newPullParser()
        parser.setInput(StringReader(xml))
        return parser
    }

    private fun parseDate(text: String): Date? {
        return try {
            dateFormat.parse(text)
        } catch (e: Exception) {
            try {
                dateFormatAlt.parse(text)
            } catch (e: Exception) {
                null
            }
        }
    }

    // Builder classes
    private class FolderBuilder {
        var serverId = ""
        var parentId = "0"
        var displayName = ""
        var type = 0

        fun build(): EASFolder? {
            if (serverId.isEmpty()) return null
            return EASFolder(serverId, parentId, displayName, type)
        }
    }

    private class EventBuilder {
        var id = ""
        var subject = ""
        var startTime: Date? = null
        var endTime: Date? = null
        var location: String? = null
        var body: String? = null
        var isAllDay = false
        var attendees = mutableListOf<EASAttendee>()
        var exceptions = mutableListOf<EASException>()

        fun build(): EASCalendarEvent? {
            val start = startTime ?: return null
            val end = endTime ?: return null
            if (id.isEmpty()) return null

            return EASCalendarEvent(
                id = id,
                subject = subject.ifEmpty { "Untitled" },
                startTimeMillis = start.time,
                endTimeMillis = end.time,
                location = location,
                body = body,
                organizer = null,
                attendees = attendees.toList(),
                isAllDay = isAllDay,
                exceptions = exceptions.toList().ifEmpty { null }
            )
        }
    }

    private class ExceptionBuilder {
        var originalStartTime: Date? = null
        var startTime: Date? = null
        var endTime: Date? = null
        var subject: String? = null
        var location: String? = null
        var isDeleted = false

        fun build(): EASException? {
            val originalStart = originalStartTime ?: return null
            return EASException(
                originalStartTimeMillis = originalStart.time,
                startTimeMillis = if (isDeleted) null else startTime?.time,
                endTimeMillis = if (isDeleted) null else endTime?.time,
                subject = subject,
                location = location,
                isDeleted = isDeleted
            )
        }
    }

    private class AttendeeBuilder {
        var email = ""
        var name = ""
        var type = 1
        var status: Int? = null

        fun build(): EASAttendee? {
            if (email.isEmpty()) return null
            return EASAttendee(
                email = email,
                name = name.ifEmpty { email },
                type = AttendeeType.fromValue(type),
                status = status?.let { ResponseStatus.fromValue(it) }
            )
        }
    }
}
